### PR TITLE
Closes issue #19. Support multiple Docker annotations with networks.

### PR DESCRIPTION
**Pull request by hanleyt**

### CODE REVIEW

File: src/main/java/com/github/junit5docker/ContainerInfo.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package com.github.junit5docker;
+
+import java.util.Collection;
+
+import static java.util.Collections.emptySet;
+import static java.util.Collections.unmodifiableCollection;
+
+class ContainerInfo {
```
Comment (FaustXVI) [inline src/main/java/com/github/junit5docker/ContainerInfo.java:8]: Thanks for fixing my primitive obsession on the container ID :) 👍

File: src/main/java/com/github/junit5docker/DefaultDockerClient.java [hunk 2]
```diff
@@ -25,12 +30,67 @@
...
+    private Optional<String> getExistingNetworkId(String networkName) {
+        return dockerClient.listNetworksCmd()
+                           .exec()
+                           .stream()
+                           .filter(name -> name.getName().equals(networkName))
+                           .reduce((name1, name2) -> {
+                               String msg = String.format("Multiple networks found with the same name: %s and %s",
+                                                          name1,
+                                                          name2);
+                               throw new IllegalStateException(msg);
```
Comment (FaustXVI) [inline src/main/java/com/github/junit5docker/DefaultDockerClient.java:76]: Is it really something that can happen or is it a docker bug ?
If it's a docker bug, I think we could just get the first network having the right name. What do you think ?

File: src/test/java/com/github/junit5docker/StartMultipleDockerAnnotationsIT.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import static com.github.junit5docker.StartMultipleDockerAnnotationsIT.TEST_NETWORK_2;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Dockers({
+    @Docker(image = "faustxvi/simple-two-ports", ports = @Port(exposed = 8081, inner = 8080),
+        networks = {TEST_NETWORK_1, TEST_NETWORK_2}),
+    @Docker(image = "faustxvi/simple-two-ports", ports = @Port(exposed = 8082, inner = 8080),
+        networks = {TEST_NETWORK_1, TEST_NETWORK_2})})
+@DisplayName("Multiple docker containers with multiple networks")
+class StartMultipleDockerAnnotationsIT {
```
Comment (FaustXVI) [inline src/test/java/com/github/junit5docker/StartMultipleDockerAnnotationsIT.java:33]: Thanks a lot for that tests !

We also have the `DefaultDockerClientIT` that you way want to add tests to (for the network). It will allow us to detect misusage of the Docker-java framework more easily (already helped us a lot :) )

Another thing we like to do it to add scenarios to the `index.feature` file. It allows you to document your feature in a more natural way and is used to generate the website (so you won't repeat yourself and the website is always up-to-date).

By the way, it looks to me like you're adding two features : multiple containers and network. I agree that the network doesn't make sens without the multiple container feature. But I think having multiple containers without (shared) networks still makes sense. What do you think ?

File: src/main/java/com/github/junit5docker/DockerExtension.java [hunk 4]
```diff
@@ -109,14 +88,29 @@ private String findImageName(Docker dockerAnnotation) {
...
     }
+
+    private void stopAndRemove(ExtensionContext context, Docker docker) {
+        ContainerInfo containerInfo = getStore(context).remove(docker, ContainerInfo.class);
+
+        String containerId = containerInfo.getContainerId();
+
+        containerInfo.getNetworkIds().forEach(networkId -> {
+            dockerClient.disconnectFromNetwork(containerId, networkId);
+            dockerClient.maybeRemoveNetwork(networkId);
```
Comment (FaustXVI) [inline src/main/java/com/github/junit5docker/DockerExtension.java:107]: The goal of the dockerClient interface is to define the interface we'd like to have with docker.

Since we want to clean up behind us, I would argue that we want to disconnect from all networks and (always?) remove them. If so, the Docker extension doesn't need to know the `networkId`s. The `DockerClient` can do it for us. 

I think that `stopAndRemoveContainer` is the perfect place to do so. Maybe we should rename it to `stopAndCleanContainer` ?

File: src/main/java/com/github/junit5docker/DockerExtension.java [hunk 5]
```diff
@@ -6,90 +6,69 @@
...
-import static java.util.concurrent.CompletableFuture.supplyAsync;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
 
 class DockerExtension implements BeforeAllCallback, AfterAllCallback, BeforeEachCallback, AfterEachCallback {
 
     private final DockerClientAdapter dockerClient;
 
-    private String containerId;
+    private final DockerLogChecker dockerLogChecker;
```
Comment (FaustXVI) [inline src/main/java/com/github/junit5docker/DockerExtension.java:19]: Awesome refactoring ! Makes the code way simpler :) 👍

File: src/main/java/com/github/junit5docker/DockerExtension.java [hunk 6]
```diff
@@ -109,14 +88,29 @@ private String findImageName(Docker dockerAnnotation) {
...
+
+        containerInfo.getNetworkIds().forEach(networkId -> {
+            dockerClient.disconnectFromNetwork(containerId, networkId);
+            dockerClient.maybeRemoveNetwork(networkId);
+        });
+        dockerClient.stopAndRemoveContainer(containerId);
+    }
+
+    private static ExtensionContext.Store getStore(ExtensionContext context) {
+        return context.getStore(ExtensionContext.Namespace.GLOBAL);
```
Comment (FaustXVI) [inline src/main/java/com/github/junit5docker/DockerExtension.java:113]: I didn't know about the store. Looks ok to me but I don't understand why you prefered this vs having a field in the class ?

File: src/main/java/com/github/junit5docker/Docker.java [hunk 7]
```diff
@@ -20,6 +21,7 @@
 @Target({ElementType.TYPE})
 @Retention(RetentionPolicy.RUNTIME)
 @ExtendWith(DockerExtension.class)
+@Repeatable(value = Dockers.class)
```
Comment (FaustXVI) [inline src/main/java/com/github/junit5docker/Docker.java:24]: If reapeted, don't we risk to call the `DockerExtension` several times ? Would removing the `ExtendWith` fix the problem (since you added it to the `Dockers` annotation)?

File: src/test/java/com/github/junit5docker/DefaultDockerClientIT.java [hunk 8]
```diff
@@ -90,7 +89,8 @@ public void ensureImageIsPulled() {
             @Test
             @DisplayName("start a container without ports")
             public void shouldStartContainer() {
-                String containerId = defaultDockerClient.startContainer(WANTED_IMAGE, emptyMap());
+                String containerId = defaultDockerClient.startContainer(WANTED_IMAGE, emptyMap(), null)
```
Comment (FaustXVI) [inline src/test/java/com/github/junit5docker/DefaultDockerClientIT.java:92]: Could we avoid using `null` ? An empty array would be fine for me (declared in constants ?)

File: src/main/java/com/github/junit5docker/ContainerInfo.java (hunk 1 above)
Comment (hanleyt) [reply src/main/java/com/github/junit5docker/ContainerInfo.java:8]: :)

File: src/main/java/com/github/junit5docker/DefaultDockerClient.java (hunk 2 above)
Comment (hanleyt) [reply src/main/java/com/github/junit5docker/DefaultDockerClient.java:76]: Hmm for some reason i thought i saw this happening during development when the networks weren't being cleaned up correctly so I added this guard. 
But testing it again now it seems you're right, that you can't have multiple networks with the same name. So i'll change it back to find first :)

File: src/main/java/com/github/junit5docker/Docker.java (hunk 7 above)
Comment (hanleyt) [reply src/main/java/com/github/junit5docker/Docker.java:24]: From the docs it seems that registering an extension a second time is fine, that the second attempt will be ignored.
http://junit.org/junit5/docs/current/user-guide/#extensions-registration-inheritance

By keeping it this way we allow users to still use the single Docker annotation by itself, without always having to wrap it in the Dockers annotation.

File: src/main/java/com/github/junit5docker/DockerExtension.java (hunk 5 above)
Comment (hanleyt) [reply src/main/java/com/github/junit5docker/DockerExtension.java:19]: thanks! it was motivated by checkstyle/pmd saying the class had too many dependencies :)

File: src/main/java/com/github/junit5docker/DockerExtension.java (hunk 4 above)
Comment (hanleyt) [reply src/main/java/com/github/junit5docker/DockerExtension.java:107]: Yes, good idea! that makes sense and would be much cleaner

File: src/main/java/com/github/junit5docker/DefaultDockerClient.java (hunk 2 above)
Comment (vdemeester) [reply src/main/java/com/github/junit5docker/DefaultDockerClient.java:76]: @hanleyt @FaustXVI I think there is a case you can if they don't have the same scope — i.e. one is local (bridge) and the other would be swarm (overaly — created elsewhere on the cluster). That said, I think it's an edge case and shouldn't really happen that much 😓

File: src/test/java/com/github/junit5docker/DefaultDockerClientIT.java (hunk 8 above)
Comment (hanleyt) [reply src/test/java/com/github/junit5docker/DefaultDockerClientIT.java:92]: 👍

File: src/test/java/com/github/junit5docker/StartMultipleDockerAnnotationsIT.java (hunk 3 above)
Comment (hanleyt) [reply src/test/java/com/github/junit5docker/StartMultipleDockerAnnotationsIT.java:33]: yes, agreed :)
I'll get working on those

File: src/main/java/com/github/junit5docker/DockerExtension.java (hunk 6 above)
Comment (hanleyt) [reply src/main/java/com/github/junit5docker/DockerExtension.java:113]: From what I understand its the recommended way to store state in extensions, and that extension classes should be stateless. It is essentially just a namespaced, hierarchical map, and having a map as a class field I imagine would perform in the exact same way. 
But apparently there are no guarantees about the exension lifecycle, so its better to use the store to store state. Plus it could be used by other extensions if they wanted to access the test state. 

https://blog.codefx.org/design/architecture/junit-5-extension-model/#Stateless

Comment (FaustXVI) [reply src/main/java/com/github/junit5docker/DockerExtension.java:113]: Awesome ! Good catch !

File: src/main/java/com/github/junit5docker/Docker.java (hunk 7 above)
Comment (FaustXVI) [reply src/main/java/com/github/junit5docker/Docker.java:24]: Great ! Let's go for it then :)

File: src/main/java/com/github/junit5docker/DefaultDockerClient.java (hunk 2 above)
Comment (FaustXVI) [reply src/main/java/com/github/junit5docker/DefaultDockerClient.java:76]: @vdemeester well, I think it's the kind of edge case we actually don't want to handle. I think that when you start to have this kind of configuration, you're not really on a development machine (dev or ci). So maybe we should actually crash at the network creation time if the network already exists. What do you think ? or we can avoid the crash by appending a random value to it ?

File: src/main/java/com/github/junit5docker/ContainerInfo.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    private final String containerId;
+
+    private final Collection<String> networkIds;
+
+    ContainerInfo(String containerId, Collection<String> networkIds) {
+        this.containerId = containerId;
+        this.networkIds = networkIds == null ? emptySet() : unmodifiableCollection(networkIds);
+    }
+
+    String getContainerId() {
```
Comment (FaustXVI) [inline src/main/java/com/github/junit5docker/ContainerInfo.java:19]: oh, i forgot it last time :) Since this is just a immutable data structure, public fields or fine for me. I don't see any value in getters.
What's your opinion on this ?

Comment (hanleyt) [reply src/main/java/com/github/junit5docker/ContainerInfo.java:19]: Hmm I've always done this out of habit after reading effective java. But on further inspection its fine to access fields directly if the class isn't public, and even more so if the fields are immutable.
So sure, I'll make that change. Thanks! :)

File: src/main/java/com/github/junit5docker/DefaultDockerClient.java (hunk 2 above)
Comment (vdemeester) [reply src/main/java/com/github/junit5docker/DefaultDockerClient.java:76]: Both would work (i.e. let the thing crash at network creation, *or* adding a random value to make sure it's unique)

**Comment by FaustXVI** [comment]
@hanleyt hi there :)
How are you ?
Any news on his PR ? That's awesome work ! Almost there ! Do you need any help ?
Regards,
